challenges/cloud/sre.md

```diff
@@ -0,0 +1,376 @@
+
+* We expect the design document to be completed roughly within the first week.
+  This is to ensure you have enough time to work on the implementation.
+* Avoid writing an overly detailed design document. Two to three pages is
+  sufficient.
+* Avoid sending us draft design documents, it is difficult to evaluate what
+  parts are draft and which parts are complete. Instead we encourage asking
+  questions in Slack and sharing a design document that is ready to be
+  reviewed.
+
+## Server & Automation
+
+Once the design document is approved, begin working on adding features to the
+server implementation that includes automation. A key feature of this challenge 
+is to produce a self-contained GitHub project that automates away as much as 
+possible. This should include code quality, testing, build, and deployment.
+
+Expect the interview panel to clone the repository and execute one or more 
+Make targets that build and test a working solution. 
+
+Add a couple of high quality tests that cover happy and unhappy scenarios. Keep
+external dependencies as low or clear as possible. Expect the reviewers to
+checkout the repository and execute the tests themselves.
+
+Do not try to achieve full test coverage. This will take too long. Write enough
+to exercise the different key components and show they are working as
+intended, as well as to demonstrate your approach to automation.
+
+## Deployment
+
+Levels 3+ include a deployment step which should deploy the solution to a local
+Kubernetes cluster. The choice of which Kubernetes cluster is up to you, but 
+please choose one that has the ability to run on macOS and Linux.
+
+## Key Dependencies
+
+You are welcome to solve this challenge using tools you are most familiar with,
+but, at a minimum, please include the following:
+
+* Go
+* make
+* Docker
+
+## Level 1
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+
+## Level 2
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of the Kubernetes Deployment
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+
+## Level 3
+
+### Server
+
+* HTTP API to retrieve the replica count of the Kubernetes Deployment
+* HTTP API to set the replica count of a the Kubernetes Deployment
+* HTTP API to get the list of available Deployments in the Kubernetes cluster
+* HTTP health check verifying Kubernetes connectivity
+
+### Automation
+
+* Ability to build the server in a Docker container
+* Ability to run the server
+* Ability to execute integration tests against the local Kubernetes cluster
+* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
```

Comment:
> all tests pass

Does this refer to the integration tests from earlier? Are there expected to be any other tests? It'd be worth clarifying to `the integration tests pass` since integration tests in CI is a much bigger lift than `go test ./...` unit tests.

Reply:
Yes, there is an expectation to write more tests:

> Add a couple of high quality tests that cover happy and unhappy scenarios. Keep
external dependencies as low or clear as possible. Expect the reviewers to
checkout the repository and execute the tests themselves.

This isn't called out as a line item for each of the levels. So I'll add it there.

There's no expectation for a lot of test coverage, (1 or 2 tests should be okay) but expectations on the quality would increase.

Reply:
Thanks for the clarification. Can we make sure to call out integration testing as explicitly in scope or out of scope for github actions?  I'd recommend:

* unit testing & static analysis/linting is in scope for CI
* integration testing is out of scope for CI

This is mostly to keep the challenge lighter.

Reply:
How about just adding:

* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
  * Unit testing and static analysis is in scope
  * Integration testing can be omitted/skipped

Reply:
> How about just adding:
> 
>     * GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
>       
>       * Unit testing and static analysis is in scope
>       * Integration testing can be omitted/skipped

Sounds good to me!

Reply:
Added - thanks!